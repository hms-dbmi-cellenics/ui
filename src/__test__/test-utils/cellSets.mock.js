const mockCellSets = {
  initialLoadPending: false,
  error: false,
  loading: false,
  accessible: true,
  hierarchy: [
    {
      key: 'louvain',
      children: [{ key: 'cluster-a' }, { key: 'cluster-b' }, { key: 'cluster-c' }],
      cellIds: new Set(),
    },
    {
      key: 'sample',
      children: [{ key: 'sample-1' }, { key: 'sample-2' }],
      cellIds: new Set(),
    },
    {
      key: 'scratchpad',
      children: [{ key: 'scratchpad-a' }],
      cellIds: new Set(),
    },
  ],
  properties: {
    'cluster-a': {
      name: 'cluster a',
      key: 'cluster-a',
      cellIds: new Set([0, 1, 6]),
      color: '#01FFFF',
    },
    'cluster-b': {
      name: 'cluster b',
      key: 'cluster-b',
      cellIds: new Set([2, 3]),
      color: '#23FFFF',
    },
    'cluster-c': {
      name: 'cluster c',
      key: 'cluster-c',
      cellIds: new Set([4, 5]),
      color: '#45FFFF',
    },
    'sample-1': {
      name: 'Sample 1',
      key: 'sample-1',
      cellIds: new Set([0, 1, 2]),
      color: '#012FFF',
    },
    'sample-2': {
      name: 'Sample 2',
      key: 'sample-2',
      cellIds: new Set([3, 4, 5]),
      color: '#345FFF',
    },
    'scratchpad-a': {
      cellIds: new Set(['5']),
      key: 'scratchpad-a',
      name: 'New Cluster',
      color: '#5FFFFF',
    },
    louvain: {
      cellIds: new Set(),
      name: 'Louvain clusters',
      key: 'louvain',
      type: 'cellSets',
      rootNode: true,
    },
    sample: {
      cellIds: new Set([6, 7, 8]),
      name: 'Samples',
      key: 'sample',
      type: 'metadataCategorical',
      rootNode: true,
    },
    scratchpad: {
      cellIds: new Set(),
      name: 'Scratchpad',
      key: 'scratchpad',
      type: 'cellSets',
      rootNode: true,
    },
  },
};

export {
  // eslint-disable-next-line import/prefer-default-export
  mockCellSets,
};
