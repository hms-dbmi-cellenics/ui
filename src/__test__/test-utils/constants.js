export default {
  EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperiment',
  CLONED_EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperclone',
  PROJECT_WITH_SAMPLES_UUID: '380c53b4-mock-test-8091-4c82336d6d49',
  PROJECT_WITHOUT_SAMPLE_UUID: '4761594b-mock-test-ba89-c2c812d39bb5',
  EXPERIMENT_NAME: 'Test Experiment',
  PROJECT_ID: 'test24b6-8600-test-mock-63822d2fmock',
  SAMPLE_ID: 'test9188-d682-test-mock-cb6d644cmock',
  API_ENDPOINT: /^http?:\/\/localhost:3000.*$/,
  S3_ENDPOINT: 'http://mock.s3.amazonaws.com',
  // pragma: allowlist secret
  JWT_TOKEN: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6Ikpfake.eyJzdWIiOiIxMjM0NTY3ODkwIiwifakeZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTfakeM5MDIyfQ.fakexwRfakeKKF2QT4fwpMeJf36POk6yJV_adQsfake',
  MOCK_DATETIME: '0000-00-00T00:00:00.000Z',
  ETAG: '380c53b4-mock-etag-8091-4c82336d6d49',
};
