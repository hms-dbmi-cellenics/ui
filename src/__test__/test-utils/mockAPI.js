import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

import fake from '__test__/test-utils/constants';

import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (
  mockedResults,
) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }

  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${experimentId}`]: () => statusResponse(200, 'OK'),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
};
