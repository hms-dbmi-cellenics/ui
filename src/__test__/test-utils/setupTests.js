import '@testing-library/jest-dom/extend-expect';
import preloadAll from 'jest-next-dynamic';

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

// This is needed, because the unit tests don't register the Vitessce imports
jest.mock('components/data-exploration/DynamicVitessceWrappers', () => ({
  Scatterplot: () => 'Mocked Scatterplot',
  Heatmap: () => 'Mocked Heatmap',
  Spatial: () => 'Mocked Spatial',
}));

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});
