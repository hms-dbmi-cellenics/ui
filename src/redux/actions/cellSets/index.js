import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import deleteCellClass from './deleteCellClass';
import runCellSetsClustering from './runCellSetsClustering';
import runCellSetsAnnotation from './runCellSetsAnnotation';
import runCellCycleScoring from './runCellCycleScoring';
import updateCellSetsClustering from './updateCellSetsClustering';

import loadCellSets from './loadCellSets';
import reorderCellSet from './reorderCellSet';
import updateCellSetProperty from './updateCellSetProperty';
import updateCellSetSelected from './updateCellSetSelected';
import setCellSetHiddenStatus from './setCellSetHiddenStatus';
import unhideAllCellSets from './unhideAllCellSets';

export {
  createCellSet,
  deleteCellSet,
  deleteCellClass,
  runCellSetsClustering,
  runCellSetsAnnotation,
  updateCellSetsClustering,
  loadCellSets,
  reorderCellSet,
  updateCellSetSelected,
  updateCellSetProperty,
  setCellSetHiddenStatus,
  unhideAllCellSets,
  runCellCycleScoring,
};
