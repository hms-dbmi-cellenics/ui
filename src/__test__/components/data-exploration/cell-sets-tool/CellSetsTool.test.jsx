import React from 'react';

import _ from 'lodash';
import {
  render, screen, fireEvent, waitFor, within,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { act } from 'react-dom/test-utils';

import { Provider } from 'react-redux';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import createTestComponentFactory from '__test__/test-utils/testComponentFactory';
import { makeStore } from 'redux/store';

import CellSetsTool from 'components/data-exploration/cell-sets-tool/CellSetsTool';
import { createCellSet } from 'redux/actions/cellSets';

import fetchWork from 'utils/work/fetchWork';

import mockAPI, { generateDefaultMockAPIResponses, promiseResponse } from '__test__/test-utils/mockAPI';
import { loadBackendStatus } from 'redux/actions/backendStatus';

enableFetchMocks();

jest.mock('utils/work/fetchWork');

jest.mock('utils/socketConnection', () => {
  const mockEmit = jest.fn();
  const mockOn = jest.fn();

  return {
    __esModule: true,
    default: new Promise((resolve) => {
      resolve({ emit: mockEmit, on: mockOn, id: '5678' });
    }),
    mockEmit,
    mockOn,
  };
});

const mockNavigateTo = jest.fn();

jest.mock('utils/AppRouteProvider', () => ({
  useAppRouter: jest.fn(() => ({
    navigateTo: mockNavigateTo,
  })),
}));

const cellSetsData = require('__test__/data/cell_sets.json');

const cellSetsWithAnnotatedCellClass = _.cloneDeep(cellSetsData);
cellSetsWithAnnotatedCellClass.cellSets.push(
  {
    key: 'annotated-cell-class-key',
    name: 'Annotated cell class',
    rootNode: true,
    children: [
      {
        key: 'annotated-cell-set-key',
        name: 'Annotated cell set',
        color: '#8c564b',
        rootNode: false,
        cellIds: [
          1, 2, 3, 4, 5,
        ],
      },
    ],
    type: 'cellSets',
  },
);

const louvainClusters = cellSetsData.cellSets.find(({ key }) => key === 'louvain').children;

const experimentId = '1234';

const getClusterByName = (clusterName) => {
  const clusterKey = Object.keys(storeState.getState().cellSets.properties).filter((key) => {
    if (storeState.getState().cellSets.properties[key].name === clusterName) {
      return key;
    }

    return undefined;
  });
  return clusterKey;
};

const defaultProps = {
  experimentId,
  width: 50,
  height: 50,
};

const cellSetsToolFactory = createTestComponentFactory(CellSetsTool, defaultProps);

let storeState;

// Mocking samples update / delete routes
const customResponses = {
  [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(JSON.stringify(cellSetsData)),
};
const mockAPIResponse = _.merge(
  generateDefaultMockAPIResponses(experimentId),
  customResponses,
);

describe('CellSetsTool', () => {
  beforeEach(async () => {
    jest.clearAllMocks();

    fetchMock.resetMocks();
    fetchMock.mockIf(/.*/, mockAPI(mockAPIResponse));

    storeState = makeStore();
  });

  it('renders correctly cell set tool with no clusters in custom cell sets', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    const editButtons = screen.getAllByLabelText(/Edit/);
    // There should be no delete buttons.
    const deleteButtons = screen.queryByText(/Delete/);

    expect(editButtons.length).toEqual(3);
    expect(deleteButtons).toBeNull();
  });

  it('renders correctly cell set tool with clusters in custom cell sets', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // create a new cluster:
    await act(async () => {
      storeState.dispatch(createCellSet(experimentId, 'New Cluster', '#3957ff', new Set([1, 2, 3, 4, 5])));
    });

    // expand custom cell sets tree
    const customCellSetsGroup = screen.getAllByRole('img', { name: 'down' })[1];
    userEvent.click(customCellSetsGroup);

    // There should be delete buttons for clusters under Custom cell sets.
    const deleteButtons = screen.getAllByLabelText(/Delete/);
    expect(deleteButtons.length).toEqual(1);
  });

  it('cell set operations should not render when no cell sets are selected', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    const cellSetOperations = screen.queryByLabelText(/of selected$/i);

    // There should be no operations rendered
    expect(cellSetOperations).toEqual(null);
  });

  it('cell set operations should render when cell sets are selected', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // expand the louvain clusters tree
    const louvainCLustersGroup = screen.getAllByRole('img', { name: 'down' })[0];
    userEvent.click(louvainCLustersGroup);

    // select the third louvain cluster
    const louvain3Cluster = screen.getByText('Cluster 3');
    userEvent.click(louvain3Cluster);

    const cellSetOperations = screen.getAllByLabelText(/of selected$/i);

    // There should be three operations rendered.
    expect(cellSetOperations.length).toEqual(3);

    // Uncomment test to test for the existence of the cell sets test
    // There should be a button for subsetting cellsets
    // const subsetCellSetsOperation = screen.getByLabelText(/Create new experiment from selected cellsets/i);
    // expect(subsetCellSetsOperation).toBeInTheDocument();
  });

  it('can compute a union of 2 cell sets', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // expand the louvain clusters tree
    const louvainCLustersGroup = screen.getAllByRole('img', { name: 'down' })[0];
    userEvent.click(louvainCLustersGroup);

    // select the third louvain cluster
    const louvain3Cluster = screen.getByText('Cluster 3');
    userEvent.click(louvain3Cluster);

    // select the fourth louvain cluster
    const louvain4Cluster = screen.getByText('Cluster 4');
    userEvent.click(louvain4Cluster);

    // ensure that initially we have 0 custom cell sets
    expect(screen.queryByText('New Cluster')).toBeNull();

    const unionOperation = screen.getByLabelText(/Union of selected$/i);
    userEvent.click(unionOperation);

    const saveButton = screen.getByLabelText(/Save/i);
    await act(async () => {
      fireEvent(
        saveButton,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
    });

    // expand custom cell sets tree
    const customCellSetsGroup = screen.getAllByRole('img', { name: 'down' })[1];
    userEvent.click(customCellSetsGroup);

    screen.getByText('New Cluster');
    const newClusterKey = getClusterByName('New Cluster');

    const cluster3CellIds = louvainClusters.find(({ name }) => name === 'Cluster 3').cellIds;
    const cluster4CellIds = louvainClusters.find(({ name }) => name === 'Cluster 4').cellIds;
    const unionCellIds = [...cluster3CellIds, ...cluster4CellIds];

    // test the union cellSet function
    const expectedUnion = new Set(unionCellIds);

    // get the cell ids of the new cluster that got created as the union of those clusters
    const actualUnion = storeState.getState().cellSets.properties[newClusterKey].cellIds;

    expect(actualUnion).toEqual(expectedUnion);
  });

  it('can compute an intersection of 2 cell sets', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // ensure that initially we have 0 custom cell sets
    expect(screen.queryByText('New Cluster')).toBeNull();

    // expand the louvain clusters tree
    const louvainCLustersGroup = screen.getAllByRole('img', { name: 'down' })[0];
    userEvent.click(louvainCLustersGroup);

    // select the first louvain cluster
    const louvain0Cluster = screen.getByText('Cluster 0');
    userEvent.click(louvain0Cluster);

    // create a new cluster with some cells that will overlap:
    await act(async () => {
      storeState.dispatch(createCellSet(experimentId, 'test cluster', '#3957ff', new Set([1, 2, 3, 4, 31, 32])));
    });

    // expand custom cell sets tree
    const customCellSetsGroup = screen.getAllByRole('img', { name: 'down' })[1];
    userEvent.click(customCellSetsGroup);

    // select the newly created cluster
    const scratchpadCluster = screen.getByText('test cluster');
    userEvent.click(scratchpadCluster);

    const intersectOperation = screen.getByLabelText(/Intersection of selected$/i);
    userEvent.click(intersectOperation);

    const saveButton = screen.getByLabelText(/Save/i);
    await act(async () => {
      fireEvent(
        saveButton,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
    });

    // expand custom cell sets tree
    const customCellSetsGroup2 = screen.getAllByRole('img', { name: 'down' })[0];
    userEvent.click(customCellSetsGroup2);

    screen.getByText('New Cluster');

    const newClusterKey = getClusterByName('New Cluster');
    const actualIntersection = storeState.getState().cellSets.properties[newClusterKey].cellIds;

    const expectedIntersection = new Set([1, 2, 3, 4]);
    expect(actualIntersection).toEqual(expectedIntersection);
  });

  it('New cluster is not created when cancel is clicked', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // ensure that initially we have 0 custom cell sets
    expect(screen.queryByText('New Cluster')).toBeNull();

    // expand the louvain clusters tree
    const louvainCLustersGroup = screen.getAllByRole('img', { name: 'down' })[0];
    userEvent.click(louvainCLustersGroup);

    // select the first louvain cluster
    const louvain0Cluster = screen.getByText('Cluster 0');
    userEvent.click(louvain0Cluster);

    // create a new cluster with some cells that will overlap:
    await act(async () => {
      storeState.dispatch(createCellSet(experimentId, 'test cluster', '#3957ff', new Set([1, 2, 3, 4])));
    });

    // expand custom cell sets tree
    const customCellSetsGroup = screen.getAllByRole('img', { name: 'down' })[1];
    userEvent.click(customCellSetsGroup);

    // select the newly created cluster
    const scratchpadCluster = screen.getByText('test cluster');
    userEvent.click(scratchpadCluster);

    const intersectOperation = screen.getByLabelText(/Intersection of selected$/i);
    userEvent.click(intersectOperation);

    const cancelButton = screen.getByLabelText(/Cancel/i);
    await act(async () => {
      fireEvent(
        cancelButton,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
    });
    expect(screen.queryByText('New Cluster')).toBeNull();
  });

  it('New cluster is not created if it will contain no cells', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // expand the louvain clusters tree
    const louvainCLustersGroup = screen.getAllByRole('img', { name: 'down' })[0];
    userEvent.click(louvainCLustersGroup);

    // select the first louvain cluster
    const louvain0Cluster = screen.getByText('Cluster 0');
    userEvent.click(louvain0Cluster);

    // select the second louvain cluster
    const louvain1Cluster = screen.getByText('Cluster 1');
    userEvent.click(louvain1Cluster);

    // ensure that initially we have 0 custom cell sets
    expect(screen.queryByText('New Cluster')).toBeNull();

    const intersectOperation = screen.getByLabelText(/Intersection of selected$/i);
    userEvent.click(intersectOperation);

    const saveButton = screen.getByLabelText(/Save/i);
    await act(async () => {
      fireEvent(
        saveButton,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
    });

    expect(screen.queryByText('New Cluster')).toBeNull();
  });

  it('cell set operations should work appropriately for complement', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // expand the louvain clusters tree
    const louvainCLustersGroup = screen.getAllByRole('img', { name: 'down' })[0];
    userEvent.click(louvainCLustersGroup);

    // select the first louvain cluster
    const louvain0Cluster = screen.getByText('Cluster 0');
    userEvent.click(louvain0Cluster);

    // compute the complement
    const complementOperation = screen.getByLabelText(/Complement of selected$/i);
    userEvent.click(complementOperation);

    // save the new cluster
    const saveButton = screen.getByLabelText(/Save/i);
    await act(async () => {
      fireEvent(
        saveButton,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
    });

    // expand custom cell sets tree
    const customCellSetsGroup = screen.getAllByRole('img', { name: 'down' })[1];
    userEvent.click(customCellSetsGroup);

    screen.getByText('New Cluster');
    const newClusterKey = getClusterByName('New Cluster');
    const actualComplement = storeState.getState().cellSets.properties[newClusterKey].cellIds;

    expect(actualComplement).toMatchSnapshot();
  });

  it('Scratchpad cluster deletion works ', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // create a new cluster:
    await act(async () => {
      storeState.dispatch(createCellSet(experimentId, 'New Cluster', '#3957ff', new Set([1, 2, 3, 4, 5])));
    });

    let cellCetsGroups = screen.getAllByRole('img', { name: 'down' });

    // expand custom cell sets tree
    userEvent.click(cellCetsGroups[1]);

    screen.getByText('New Cluster');
    const newClusterKey = getClusterByName('New Cluster');

    expect(cellCetsGroups.length).toEqual(4);

    let isInRedux = Object.keys(
      storeState.getState().cellSets.properties,
    ).includes(newClusterKey[0]);

    expect(isInRedux).toBe(true);

    // There should be a delete button for the scratchpad cluster.
    // const deleteButton = screen.getByLabelText(/Delete/);
    const deleteButtons = screen.getAllByLabelText(/Delete/);
    expect(deleteButtons.length).toEqual(1);

    // Clicking on one of the buttons...
    await act(async () => {
      fireEvent(
        deleteButtons[0],
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
    });

    // get all the cell set groups
    cellCetsGroups = screen.getAllByRole('img', { name: 'down' });

    await waitFor(() => {
      // This test used to assert that "New Cluster" text is not found in "screen"
      // in order to verify that deletion was successful.
      // Due to behaviour that I couldn't possibly explain, "New Cluster" is still
      // found in "screen" when it shouldn't. As a workaround, now the following lines test that:
      // -- the Redux store is updated successfully -- "New Cluster" should not be there
      // -- the "Custom cell sets" tree cannot be expanded.
      //    This means that is has no children, and hence "New Cluster" is deleted.
      expect(cellCetsGroups.length).toEqual(3);

      isInRedux = Object.keys(storeState.getState().cellSets.properties).includes(newClusterKey[0]);
      expect(isInRedux).toBe(false);
    });
  });

  it('calculates filtered cell indices correctly', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // expand samples tree
    const samplesGroup = screen.getAllByRole('img', { name: 'down' })[1];
    userEvent.click(samplesGroup);

    // select a sample cluster
    const wt1Cluster = screen.getByText('WT1');
    userEvent.click(wt1Cluster);

    // Cells in wt1 minus all the ones that are not in louvain (they were filtered out)
    screen.getByText('28 cells selected');
  });

  it('Displays a cell set hidden message when cluster is hidden', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // expand the louvain clusters tree
    const louvainCLustersGroup = screen.getAllByRole('img', { name: 'down' })[0];
    userEvent.click(louvainCLustersGroup);

    // hide the first cluster
    const hideButtons = screen.getAllByText('Hide');
    userEvent.click(hideButtons[0]);

    screen.getByText(/1 cell set is currently hidden./);
    screen.getAllByText(/Unhide all/);
  });

  it('Unhides clusters when the unhide button is presssed', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // expand the louvain clusters tree
    const louvainCLustersGroup = screen.getAllByRole('img', { name: 'down' })[0];
    userEvent.click(louvainCLustersGroup);

    // hide the first cluster
    let hideButtons = screen.getAllByText('Hide');
    userEvent.click(hideButtons[0]);

    // hide the second cluster
    hideButtons = screen.getAllByText('Hide');
    userEvent.click(hideButtons[0]);

    screen.getByText(/2 cell sets are currently hidden./);

    const unhideButton = screen.getAllByText('Unhide');
    expect(unhideButton.length).toEqual(2);

    // now unhide the first cluster
    userEvent.click(unhideButton[0]);

    screen.getByText('Unhide');
    screen.getByText(/1 cell set is currently hidden./);
    screen.getAllByText(/Unhide all/);
  });

  it('Unhides everything when the unhide all button is presssed', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // expand the louvain clusters tree
    const louvainCLustersGroup = screen.getAllByRole('img', { name: 'down' })[0];
    userEvent.click(louvainCLustersGroup);

    let hideButtons = screen.getAllByText('Hide');
    expect(hideButtons.length).toEqual(14);

    // hide the first cluster
    userEvent.click(hideButtons[0]);

    hideButtons = screen.getAllByText('Hide');
    expect(hideButtons.length).toEqual(13);

    // hide the second cluster
    userEvent.click(hideButtons[0]);

    screen.getByText(/2 cell sets are currently hidden./);

    const unhideAllButton = screen.getAllByText('Unhide all');
    expect(unhideAllButton.length).toEqual(1);

    const unhideButtons = screen.getAllByText('Unhide');
    expect(unhideButtons.length).toEqual(2);

    // now unhide all clusters
    userEvent.click(unhideAllButton[0]);

    expect(screen.queryByText(/2 cell sets are currently hidden./)).toBeNull();
    expect(screen.queryByText('Unhide')).toBeNull();
  });

  it('Runs subset experiment correctly', async () => {
    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // expand the louvain clusters tree
    const louvainCLustersGroup = screen.getAllByRole('img', { name: 'down' })[0];
    userEvent.click(louvainCLustersGroup);

    // select the third louvain cluster
    const louvain3Cluster = screen.getByText('Cluster 3');
    userEvent.click(louvain3Cluster);

    // select the fourth louvain cluster
    const louvain4Cluster = screen.getByText('Cluster 4');
    userEvent.click(louvain4Cluster);

    const subsetOperation = screen.getByLabelText('Create new experiment from selected cellsets');
    userEvent.click(subsetOperation);

    const createModal = screen.getByTestId('subsetCellSetsModal');
    const createButton = within(createModal).getByRole('button', { name: /Create/i });

    await act(async () => {
      fireEvent(
        createButton,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
    });

    expect(fetchMock).toHaveBeenCalledWith(
      expect.stringContaining(`/v2/experiments/${experimentId}/subset`),
      expect.objectContaining({
        body: expect.stringContaining('"cellSetKeys":["louvain-3","louvain-4"]'),
        method: 'POST',
      }),
    );
  });

  it('Annotated cell sets has delete buttons', async () => {
    const mockAPICellClassAPIResponse = {
      ...mockAPIResponse,
      [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(JSON.stringify(cellSetsWithAnnotatedCellClass)),
    };

    fetchMock.mockIf(/.*/, mockAPI(mockAPICellClassAPIResponse));

    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // The annotated cell clas is the last class in the hierarchy
    const annotatedCellSetsGroup = screen.getAllByRole('img', { name: 'down' }).pop();
    userEvent.click(annotatedCellSetsGroup);

    // There should only be one delete button for the annotated cell class
    const deleteButtons = screen.getAllByLabelText(/Delete/);
    expect(deleteButtons.length).toEqual(1);
  });

  it('Deleting annotated cell class deletes the cell class and the cell sets', async () => {
    const mockAPICellClassAPIResponse = {
      ...mockAPIResponse,
      [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(JSON.stringify(cellSetsWithAnnotatedCellClass)),
    };

    fetchMock.mockIf(/.*/, mockAPI(mockAPICellClassAPIResponse));

    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // The annotated cell clas is the last class in the hierarchy
    const deleteAnnotatedCellClassButton = screen.getByLabelText(/Delete/);
    act(() => {
      userEvent.click(deleteAnnotatedCellClassButton);
    });

    // get all the cell set groups
    const numCellGroupsAfterDelete = screen.getAllByRole('img', { name: 'down' }).length;

    await waitFor(() => {
      // This test used to assert that "Annotated cell set" text is not found in "screen"
      // in order to verify that deletion was successful.
      // 4 Cell groups: louvain, custom cell sets, sample, Track_1 (metadata)
      expect(numCellGroupsAfterDelete).toEqual(4);

      expect(screen.queryByText('Annotated cell class')).toBeNull();

      // Check that cell class does not exist anymore in redux
      const isClassInRedux = Object.keys(storeState.getState().cellSets.properties).includes('annotated-cell-class-key');
      expect(isClassInRedux).toBe(false);

      // Check that cell set does not exist anymore in redux
      const isCellSetInRedux = Object.keys(storeState.getState().cellSets.properties).includes('annotated-cell-set-key');
      expect(isCellSetInRedux).toBe(false);
    });
  });
});

// Enable when AnnotateClustersTool panel is enabled
describe('AnnotateClustersTool', () => {
  beforeEach(async () => {
    jest.clearAllMocks();

    await storeState.dispatch(loadBackendStatus(experimentId));

    await act(async () => {
      render(
        <Provider store={storeState}>
          {cellSetsToolFactory()}
        </Provider>,
      );
    });

    // Switch to tab
    const annotateClustersTabTitle = screen.getByText('Annotate clusters');

    // Switch to tab
    userEvent.click(annotateClustersTabTitle);
  });

  it('Renders correctly', async () => {
    // Check displays correct text
    screen.getByText(/ScType/);
    screen.getByText(/Tissue Type/);
    screen.getByText(/Species/);
    screen.getByText(/Compute/);

    // Displays correct placeholder in selects
    screen.getByText('Select a tissue type');
    screen.getByText('Select a species');

    // Displays button and it's disabled
    expect(screen.getByRole('button', { name: /Compute/ })).toBeDisabled();
  });

  it('Can dispatch work request', async () => {
    const tissueSelector = screen.getAllByRole('combobox')[0];
    const speciesSelector = screen.getAllByRole('combobox')[1];

    // Choose the liver option
    userEvent.click(tissueSelector);
    userEvent.click(screen.getByText(/Liver/));

    // Choose the mouse option
    userEvent.click(speciesSelector);
    userEvent.click(screen.getAllByText(/mouse/)[1]);

    // Now the button's enabled
    const button = screen.getByRole('button', { name: /Compute/ });
    expect(button).toBeEnabled();

    // Click the button
    act(() => {
      userEvent.click(button);
    });

    // It dispatches a work request
    await waitFor(() => {
      expect(fetchWork).toMatchSnapshot();
    });
  });
});
