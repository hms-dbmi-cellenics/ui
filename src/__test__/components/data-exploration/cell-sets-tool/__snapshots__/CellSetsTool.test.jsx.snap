// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnotateClustersTool Can dispatch work request 1`] = `
[MockFunction] {
  "calls": [
    [
      "1234",
      {
        "name": "ScTypeAnnotate",
        "species": "mouse",
        "tissue": "Liver",
      },
      [Function],
      [Function],
      {
        "broadcast": true,
        "timeout": 180.172,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CellSetsTool cell set operations should work appropriately for complement 1`] = `
Set {
  31,
  32,
  33,
  34,
  35,
  36,
  37,
  38,
  39,
  40,
  41,
  42,
  43,
  47,
  48,
  49,
  50,
  51,
  52,
  53,
  54,
  55,
  57,
  58,
  59,
  60,
  61,
  62,
  64,
  65,
  66,
  68,
  69,
  70,
  71,
  72,
  73,
  75,
  76,
  77,
  79,
  80,
  81,
  82,
  84,
  85,
  88,
  89,
  90,
  91,
  92,
  93,
  94,
  95,
  96,
  97,
  98,
  99,
  100,
}
`;

exports[`ScoreCellCycleTool Can dispatch work request 1`] = `
[MockFunction] {
  "calls": [
    [
      "1234",
      {
        "name": "CellCycleScoring",
      },
      [Function],
      [Function],
      {
        "broadcast": true,
        "timeout": 180.172,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
